server/routes/controllers/forgotPasswordController.js

```diff
@@ -20,23 +23,56 @@ const forgotPasswordController = {
     return Math.floor(Math.random() * 900000) + 100000 // generate a 6 digit code
   },
 
+  _getUser (email) {
+    return this.Users.findOne({email: email}).exec((err, forgot) => {
+      if (err) {
+        return false
+      } else {
+        return true
+      }
+    })
+  },
+
   sendVerificationCodeEmail (req, res) {
     const email = req.body.email
-    const code = this._generateCode()
-    this._saveIntoDatabase(email, code)
-      .then(saveForgot => {
-        const text = `Your validation code is: ${code}`
-        const message = {
-          from: 'team.calltocode@gmail.com',
-          to: email,
-          subject: `Validation code for ${email}`,
-          html: `<strong>${text}</strong>`,
-          text
-        }
-        this.emailClient.send(message)
+
+    this._getUser(email).then((exists) => {
+      // if the user does not exist, do not generate code and send email. It's expected that the user inserts a valid email.
+      if (!exists) {
         return res.status(200).send({ status: 200 })
-      })
-      .catch(error => { if (error) return res.status(500).send({ status: 500 }) })
+      }
+      const code = this._generateCode()
+      this._saveIntoDatabase(email, code)
+        .then(saveForgot => {
+          const text = `Your validation code is: ${code}`
+          const message = {
+            from: 'team.calltocode@gmail.com',
+            to: email,
+            subject: `Validation code for ${email}`,
+            html: `<strong>${text}</strong>`,
+            text
+          }
+          this.emailClient.send(message)
+          return res.status(200).send({ status: 200 })
+        })
+        .catch(error => { if (error) return res.status(500).send({ status: 500 }) })
+    }).catch(error => { if (error) return res.status(500).send({ status: 500 }) })
+  },
+
+  validateCode (req, res) {
+    const {email, code} = req.body
+    this.ForgotPass.findOne({email: email, code: code}).exec((err, forgot) => {
+      if (err) {
```

Comment:
This is quite a bit of nested conditional logic.  It could de-nested by taking advantage of the fact that no code after a `return` will be executed...

```
if (err) {
  return res.status(404).json({error: 'Invalid code or email'})
}

if (forgot) {
  forgot.remove()
  return res.status(200).json({status: 200})
}

return res.status(404).json({error: 'Invalid code or email'})
```

Reply:
true fact. Thanks.